tarpc/src/client.rs

```diff
@@ -764,72 +764,138 @@ mod tests {
 
         assert!(dispatch.as_mut().poll_next_request(cx).is_pending());
     }
-
     #[tokio::test]
     async fn test_client_error_handling() {
         let _ = tracing_subscriber::fmt().with_test_writer().try_init();
-        let (to_dispatch, pending_requests) = mpsc::channel(1);
-        let (cancellation, canceled_requests) = cancellations();
-        let transport = MockTransport(PhantomData);
-
-        let mut dispatch = Box::pin(RequestDispatch::<String, String, _> {
-            transport: transport.fuse(),
-            pending_requests,
-            canceled_requests,
-            in_flight_requests: InFlightRequests::default(),
-            config: Config::default(),
-        });
-        let mut channel = Channel {
-            to_dispatch,
-            cancellation,
-            next_request_id: Arc::new(AtomicUsize::new(0)),
-        };
-
-        let cx = &mut Context::from_waker(noop_waker_ref());
-        let (tx, mut rx) = oneshot::channel();
-        let resp = send_request(&mut channel, "hi", tx, &mut rx).await;
-        assert!(dispatch.as_mut().poll(cx).is_pending());
+        for (error, cause) in vec![
```

Comment:
Ugh, I'm sorry about this, I just realized I had left my comment about test coverage on the wrong enum. This test, which is testing RpcError::Send, was totally fine as you had it before.

What I meant to say, but failed to, was that RpcError had some other variants added in this PR, and those new variants could use test coverage—are they returned in the proper circumstances? For example, Disconnected was replaced by Shutdown and Receive errors.

The additional test cases I had in mind:

- RpcError::Shutdown is returned when the dispatch task has shut down, preventing the client from:
    1. initiating an RPC
    2. completing an RPC
- An RPC should be completed with RpcError::Receive if the dispatch task receives an error when attempting to receive a message from the transport. Failing to receive is a permanent error, which shouldn't change, but the dispatch task should still send the more specific Receive error to all pending RPCs (of which there could be a few).

I've added comments to the specific places in the code that correspond to these test cases. I actually think RpcError::Receive is unused in this PR, so you can remove it if you'd like, and I can file an issue to add it back later.

Reply:
No worries @tikue , it happens. I think now I added what you had in mind. I don't think it makes sense to delete what I already added though? It seems useful for the future and more tests usually don't hurt, WDYT?

Reply:
That sounds good to me, though I think these are really 5 different test cases (should be different `#[test]` fns). It's the match expression that leads me to think that — each error variant has different test logic.

I like the test coverage, but would prefer to put the test cases in different fns.